fri/src/verifier.rs

```diff
@@ -0,0 +1,282 @@
+            first_layer_callback,
+            n_layers,
+            first_eval_point: None,
+            eval_points: vec![F::zero(); n_layers - 1],
+            table_verifiers: vec![],
+            query_indices: vec![],
+            query_results: vec![],
+            expected_last_layer: None,
+            eval_points_test,
+            last_layer_coefficients_test,
+        }
+    }
+
+    pub fn verify_fri(&mut self) -> Result<(), Box<dyn Error>> {
+        self.commitment_phase()?;
+
+        // // query phase
+        //self.query_indices = self.choose_query_indices();
+        // //self.channel.begin_query_phase();
+
+        // // decommitment phase
+        // // TODO : annotation
+
+        // self.verify_first_layer();
+        // self.verify_inner_layers();
+        // self.verify_last_layer();
+        Ok(())
+    }
+
+    pub fn commitment_phase(&mut self) -> Result<(), Box<dyn Error>> {
+        let mut basis_index = 0;
+        for i in 0..self.n_layers {
+            let cur_fri_step = self.params.fri_step_list[i];
+            basis_index += cur_fri_step;
+
+            if i == 0 {
+                if self.params.fri_step_list[0] != 0 {
+                    self.first_eval_point = Some(self.channel.draw_felem());
+                }
+            } else {
+                self.eval_points[i - 1] = self.channel.draw_felem();
+            }
+
+            if i < self.n_layers - 1 {
+                let coset_size = 1 << self.params.fri_step_list[i + 1];
+                let n_rows = self.params.fft_domains[basis_index].size() / coset_size;
+                let n_columns = coset_size;
+                let size_of_row = ((F::MODULUS_BIT_SIZE.div_ceil(8) * 8) as usize) * n_columns;
+
+                let commitment_scheme = make_commitment_scheme_verifier(
+                    size_of_row,
+                    n_rows,
+                    0,
+                    self.commitment_hashes.clone(),
+                    n_columns,
+                );
+
+                let mut table_verifier = TableVerifier::new(n_columns, commitment_scheme);
+                let _ = table_verifier.read_commitment(&mut self.channel);
+                self.table_verifiers.push(table_verifier);
+            }
+        }
+
+        if let Some(eval_points_test) = self.eval_points_test.as_ref() {
+            self.first_eval_point = Some(eval_points_test[0]);
+            self.eval_points[..(self.n_layers - 1)]
+                .copy_from_slice(&eval_points_test[1..((self.n_layers - 1) + 1)]);
+        }
+        self.read_last_layer_coefficients()?;
+        Ok(())
+    }
+
+    pub fn read_last_layer_coefficients(&mut self) -> Result<(), Box<dyn Error>> {
+        let fri_step_sum: usize = self.params.fri_step_list.iter().sum();
+        let last_layer_size = self.params.fft_domains[fri_step_sum].size();
+
+        let mut last_layer_coefficients_vector = if let Some(last_layer_coefficients_test) =
+            self.last_layer_coefficients_test.as_ref()
```

Comment:
When does this path occur?

It seems to me that the prover would need to know the final folded polynomial before starting the FRI folding phase?